Diploma stage 4

ykuptsova opened the pull request.

ykuptsova committed:
Update README.md

ykuptsova committed:
MainApi

ykuptsova committed:
MoviesApi

ykuptsova committed:
Movies rendering from the server

ykuptsova committed:
register form validation

ykuptsova committed:
login form validation

ykuptsova committed:
profile form validation

ykuptsova committed:
Movies loading and filtering login, preloader

ykuptsova committed:
Save movie

ykuptsova committed:
pull request link

SergeyK2004 commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 

    **В вашем задании: при каждом поиске на странице фильмов результаты поиска фильмов нужно сохранять в localstorage. В сохраняемые данные должны входить и строка поиска и статус чекбокса «Короткометражки», чтобы пользователю при повторном переходе на страницу фильмов, можно было вывести на экран информацию, полностью совпадающую с его последним поиском (с расстановкой актуального состояния лайков) перед уходом на другую страницу или перезагрузкой страницы. Эту информацию можно будет доставать из хранилища при монтировании и перерисовке компоненты фильмов. При логауте пользователя нужно эту информацию в хранилище и стейте (useState, если она там есть) уничтожать.**

    **Сейчас пользователь не видит точного результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы, если на странице сохраненных использовать поиск. Поиск на разных страницах не должен зависеть друг от друга.**

	**Замечание: Сохранять данные последнего поиска на странице «Сохранённые фильмы» для их показа при повторном входе на эту страницу не нужно. Надо при входе на эту страницу выводить на экран все сохранённые фильмы, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на страницу «Сохранённые фильмы» пользователь видел все свои сохранённые карточки без задания параметров поиска.**

	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
   - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	**Комментарий: при сохранение профиля пользователь должен увидеть сообщение о подтверждении, вид сообщения может быть любым на ваш выбор**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    **Комментарий: При переходе на страницу "Сохраненные фильмы", на ней не должно быть поискового запроса со старницы "Фильмы". На каждой странице поисковый запрос свой. При входе на "Сохраненнные", данные в форме поиска должны очищаться.**
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
	**Комментарий: происходит редирект на `/signin` , не на `/`**
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используются хуки роутинга.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
	**Комментарий: Страница 404 отображается, но при нажатии на кнопку "Назад" происходит редирект на главную.**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  **Комментарий: проверки, что jwt есть в хранилище, недостаточно, при неправильном токене не происходит редирект на страницу входа**
    - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [ ] <!--1--> Цепочка обработки промисов завершается блоком catch.
	**Комментарий: блоков catch не должно быть в самих запросах, следует их расположить только в конце цепочки**
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
	**Комментарий: используется индекс массива**
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
	**Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при поиске или нажатии кнопки "Еще" на разных размерах экрана и точки перестроения экрана нужно вынести в константы**

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 51

ykuptsova committed:
bug fixes

force search on saved-movies page on mount, separate search state for saved-movies page

ykuptsova committed:
bug fixes

automatically logout user if jwt is incorrect, block profile form from editing during request

ykuptsova committed:
bug fixes

profile edit confirmation message added

ykuptsova committed:
media query upd

ykuptsova committed:
constants added

SergeyK2004 commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
   - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используются хуки роутинга.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
    - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [ ] <!--1--> Цепочка обработки промисов завершается блоком catch.
	**Комментарий: блоков catch не должно быть в самих запросах, следует их расположить только в конце цепочки**
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
	**Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 97

## Отлично, теперь необходимый функционал работает. Поздравляю с завершением дипломного проекта!


### Несколько советов для дальнейшего развития:
- сверстайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- еще немного попрактиковаться с React, например изучить redux
- будет большим плюсом при поиске работы, если немного изучите TypeScript т.к. на него переходят многие компании
- ходить на собеседования, опыт прохождения собеседований не менее важен, чем знание теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте [habr](https://habr.com/ru/) и слушайте подкаксты, они очень хорошо расширяют кругозор, и позволяют узнать, что происходит в отрасли:
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

ykuptsova commented:
Сергей, спасибо за ревью, обратную связь и рекомендации!
